src/frontend/src/utils/stream_graph_formatter.rs

```diff
@@ -93,6 +135,89 @@ impl StreamGraphFormatter {
         };
         writeln!(f, "{}{}", " ".repeat(level * 2), one_line_explain)?;
+        let explain_table_oneline = match node.get_node_body().unwrap() {
+            stream_node::NodeBody::Materialize(node) => Some(format!(
+                "materialized table: {}",
+                self.add_table(node.get_table().unwrap())
+            )),
+            stream_node::NodeBody::GlobalSimpleAgg(node) => Some(format!(
+                "state tables: [{}]",
+                node.internal_tables
+                    .iter()
+                    .map(|tb| { self.add_table(tb) })
+                    .join(", ")
+            )),
+            stream_node::NodeBody::HashAgg(node) => Some(format!(
+                "state tables: [{}]",
+                node.internal_tables
+                    .iter()
+                    .map(|tb| { self.add_table(tb) })
+                    .join(", ")
+            )),
+            stream_node::NodeBody::AppendOnlyTopN(node) => Some(format!(
+                "state table: {}",
+                self.add_table(node.get_table().unwrap())
+            )),
+            stream_node::NodeBody::HashJoin(node) => Some(format!(
+                "left table: {}, right table {},{}{}",
+                self.add_table(node.get_left_table().unwrap()),
+                self.add_table(node.get_right_table().unwrap()),
+                match &node.left_degree_table {
+                    Some(tb) => format!(" left degree table: {},", self.add_table(tb)),
+                    None => "".to_string(),
+                },
+                match &node.right_degree_table {
+                    Some(tb) => format!(" right degree table: {},", self.add_table(tb)),
+                    None => "".to_string(),
+                },
+            )),
+            stream_node::NodeBody::TopN(node) => Some(format!(
+                "state table: {}",
+                self.add_table(node.get_table().unwrap())
+            )),
+            stream_node::NodeBody::Lookup(node) => Some(format!(
+                "arrange table: {}",
+                self.add_table(node.get_arrangement_table().unwrap())
+            )),
+            stream_node::NodeBody::Arrange(node) => Some(format!(
+                "arrange table: {}",
+                self.add_table(node.get_table().unwrap())
+            )),
+            stream_node::NodeBody::DynamicFilter(node) => Some(format!(
+                "left table: {}, right table {}",
+                self.add_table(node.get_left_table().unwrap()),
+                self.add_table(node.get_right_table().unwrap()),
+            )),
+            stream_node::NodeBody::GroupTopN(node) => Some(format!(
+                "state table: {}",
+                self.add_table(node.get_table().unwrap())
+            )),
+            _ => None,
+        };
+        if let Some(explain_table_oneline) = explain_table_oneline {
+            writeln!(f, "{}{}", " ".repeat(level * 2 + 1), explain_table_oneline)?;
```

Comment:
The +1 indent looks not very clear, e.g.,

```
    Fragment 0
      StreamMaterialize { columns: [l_returnflag, l_linestatus, sum_qty, sum_base_price, sum_disc_price, sum_charge, avg_qty, avg_price, avg_disc, count_order], pk_columns: [l_returnflag, l_linestatus] }
       materialized table: 4294967294
        StreamProject { exprs: [lineitem.l_returnflag, lineitem.l_linestatus, sum(lineitem.l_quantity), sum(lineitem.l_extendedprice), sum((lineitem.l_extendedprice * (1:Int32 - lineitem.l_discount))), sum(((lineitem.l_extendedprice * (1:Int32 - lineitem.l_discount)) * (1:Int32 + lineitem.l_tax))), RoundDigit((sum(lineitem.l_quantity) / count(lineitem.l_quantity)), 4:Int32), RoundDigit((sum(lineitem.l_extendedprice) / count(lineitem.l_extendedprice)), 4:Int32), RoundDigit((sum(lineitem.l_discount) / count(lineitem.l_discount)), 4:Int32), count] }
          StreamHashAgg { group_key: [lineitem.l_returnflag, lineitem.l_linestatus], aggs: [count, sum(lineitem.l_quantity), sum(lineitem.l_extendedprice), sum((lineitem.l_extendedprice * (1:Int32 - lineitem.l_discount))), sum(((lineitem.l_extendedprice * (1:Int32 - lineitem.l_discount)) * (1:Int32 + lineitem.l_tax))), sum(lineitem.l_quantity), count(lineitem.l_quantity), sum(lineitem.l_extendedprice), count(lineitem.l_extendedprice), sum(lineitem.l_discount), count(lineitem.l_discount), count] }
           state tables: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
            StreamExchange Hash([0, 1]) from 1

```

What about

```
    Fragment 0
      StreamMaterialize { columns: [l_returnflag, l_linestatus, sum_qty, sum_base_price, sum_disc_price, sum_charge, avg_qty, avg_price, avg_disc, count_order], pk_columns: [l_returnflag, l_linestatus] }
          materialized table: 4294967294
        StreamProject { exprs: [lineitem.l_returnflag, lineitem.l_linestatus, sum(lineitem.l_quantity), sum(lineitem.l_extendedprice), sum((lineitem.l_extendedprice * (1:Int32 - lineitem.l_discount))), sum(((lineitem.l_extendedprice * (1:Int32 - lineitem.l_discount)) * (1:Int32 + lineitem.l_tax))), RoundDigit((sum(lineitem.l_quantity) / count(lineitem.l_quantity)), 4:Int32), RoundDigit((sum(lineitem.l_extendedprice) / count(lineitem.l_extendedprice)), 4:Int32), RoundDigit((sum(lineitem.l_discount) / count(lineitem.l_discount)), 4:Int32), count] }
          StreamHashAgg { group_key: [lineitem.l_returnflag, lineitem.l_linestatus], aggs: [count, sum(lineitem.l_quantity), sum(lineitem.l_extendedprice), sum((lineitem.l_extendedprice * (1:Int32 - lineitem.l_discount))), sum(((lineitem.l_extendedprice * (1:Int32 - lineitem.l_discount)) * (1:Int32 + lineitem.l_tax))), sum(lineitem.l_quantity), count(lineitem.l_quantity), sum(lineitem.l_extendedprice), count(lineitem.l_extendedprice), sum(lineitem.l_discount), count(lineitem.l_discount), count] }
              state tables: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
            StreamExchange Hash([0, 1]) from 1

```

But IMO ideally other fields should also be put in separate lines, and then state table is just like another attribute

Reply:
Sorry to asy that it is hard to do that now. because we get the attributes from the `PlanNode` but the local table id is assigned after the fragmentor.